src/librustc_trans/trans/intrinsic.rs

```diff
@@ -376,7 +378,19 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
         }
         (_, "needs_drop") => {
             let tp_ty = *substs.types.get(FnSpace, 0);
-            C_bool(ccx, type_needs_drop(ccx.tcx(), tp_ty))
+
+            // FIXME (#22815): note that both forms of type_needs_drop
+            // conservatively approximate in some cases and may say a
+            // type expression needs drop when it actually does not.
+            //
+            // This may actually matter, since some code may use the
+            // `needs_drop` intrinstic and expect it always know the
+            // exact answer. (But then again, it is not clear what the
+            // actual effect would be, apart from forcing some code to
+            // go down slower paths than necessary at runtime; e.g. in
+            // the untyped Arena allocation and deallocation paths.)
+
+            C_bool(ccx, bcx.fcx.type_needs_drop(tp_ty))
```

Comment:
We should just document that the intrinsic will always return true if there is a dtor, and *may* return true if the type does not implement `Copy` but does not actually have a dtor.